torch/csrc/jit/frontend/ir_emitter.cpp

```diff
@@ -3276,7 +3369,9 @@ struct to_ir {
     // AnyType is the only user-exposed type which we don't unify to from
     // its subtypes, so we add a cast for use cases like
     // x : Any = 1 if cond else "str"
-    if (type_hint == AnyType::get() && out_val->type() != AnyType::get()) {
+    if (type_hint &&
+          ((type_hint == AnyType::get() && out_val->type() != AnyType::get()) ||
```

Comment:
this was valid because you can always cast something to Any, it's not valid to always cast something to Union

Reply:
Changing this to "If the type hint is Union and the output is a subtype of that Union, then insert an unchecked cast". Let me know if that works